src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/IUpgradeContext.cs

```diff
@@ -30,5 +30,26 @@ public interface IUpgradeContext : IDisposable
         IDictionary<string, string> GlobalProperties { get; }
 
         ValueTask ReloadWorkspaceAsync(CancellationToken token);
+
+        /// <summary>
+        /// Returns a property value stored in the context.
+        /// </summary>
+        /// <param name="propertyName">Name identifier for the property.</param>
+        /// <returns>The property value.</returns>
+        string? TryGetPropertyValue(string propertyName);
+
+        /// <summary>
+        /// Stores a property value to the context.
+        /// </summary>
+        /// <param name="propertyName">Name identifier for the property.</param>
+        /// <param name="value">The property value to store.</param>
+        /// <param name="persistent">Whether to persist the value when the program terminates.</param>
+        void SetPropertyValue(string propertyName, string value, bool persistent);
+
+        /// <summary>
+        /// Returns persistent properties for storage.
+        /// </summary>
+        /// <returns>All defined persistent properties.</returns>
+        Dictionary<string, string> GetPersistentProperties();
```

Comment:
Let's expose this as an `IEnumerable<KeyValuePair<string,string>>`.

With 3 methods here, it may make sense to  expose it as a new interface

```cs
public interface IContextProperties:IEnumerable<KeyValuePair<string,string>>
{
  GetValue(...)
  SetValue(....)
}
```

Reply:
Ok, the refactoring was a bit bigger than I expected. I separated IContextProperties as a separate singleton service. I relocated the implementation as it has nothing to do with MSBuild now. I decided against directly inheriting from IEnumerable because I wanted to give the method a name to explain why it only returns persistent properties.

Most classes aren't really interested about whether the values are persistent or not so there could (in principle) be another interface for this but I feel this is too much complexity for a very minor advantage.

Also, added unit tests.

Reply:
Sounds good to me. The only change would be to not have it a separate service as we want the properties scoped to a context which isn't something tracked by DI (we have a factory that will create it on demand)